components/org.wso2.carbon.identity.workflow.engine/src/main/java/org/wso2/carbon/identity/workflow/engine/ApprovalTaskServiceImpl.java

```diff
@@ -0,0 +1,631 @@
+        if (ENTITY_TYPE_USERS.equals(approverDTO.getApproverType()) ||
+                ENTITY_TYPE_CLAIMED_USERS.equals(approverDTO.getApproverType())) {
+            if (approverDTO.getApproverName().equals(userId)) {
+                isAssignedApprovalTask = true;
+            }
+        } else {
+            List<String> roleIds = getAssignedRoleIds(userId, tenantDomain);
+            if (roleIds.contains(approverDTO.getApproverName())) {
+                isAssignedApprovalTask = true;
+            }
+        }
+        if (!isAssignedApprovalTask) {
+            throw new WorkflowEngineClientException(
+                    WorkflowEngineConstants.ErrorMessages.USER_ERROR_APPROVAL_TASK_IS_NOT_ASSIGNED.getCode(),
+                    WorkflowEngineConstants.ErrorMessages.USER_ERROR_APPROVAL_TASK_IS_NOT_ASSIGNED.
+                            getDescription());
+        }
+        if (WorkflowEngineConstants.TaskStatus.BLOCKED.name().equals(approvalTaskDAO.getApprovalTaskStatus(taskId))) {
+            throw new WorkflowEngineClientException(
+                    WorkflowEngineConstants.ErrorMessages.USER_ERROR_TASK_ALREADY_CLAIMED.getCode(),
+                    WorkflowEngineConstants.ErrorMessages.USER_ERROR_TASK_ALREADY_CLAIMED.getDescription());
+        }
+    }
+
+    private int numOfStates(WorkflowRequest request) throws WorkflowEngineException {
+
+        List<Parameter> parameterList = getParameterList(request);
+        int maxStep = 0;
+
+        for (Parameter parameter : parameterList) {
+            if (parameter.getParamName().equals(WorkflowEngineConstants.ParameterName.USER_AND_ROLE_STEP)) {
+                String value = parameter.getqName();
+                if (value != null && !value.isEmpty()) {
+                    String[] parts = value.split("-");
+                    if (parts.length > 1) {
+                        try {
+                            int step = Integer.parseInt(parts[1]);
+                            if (step > maxStep) {
+                                maxStep = step;
+                            }
+                        } catch (NumberFormatException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                }
+            }
+        }
+        return maxStep;
+    }
+
+    private void handleApproval(String taskId) throws WorkflowEngineServerException {
+
+        String requestID = approvalTaskDAO.getWorkflowRequestIdByApprovalTaskId(taskId);
+        List<String> taskIds = approvalTaskDAO.getApprovalTasksByWorkflowRequestId(requestID);
+        String completedStatus = WorkflowEngineConstants.TaskStatus.COMPLETED.toString();
+
+        for (String id : taskIds) {
+            approvalTaskDAO.updateApprovalTaskStatus(id, completedStatus);
+        }
+    }
+
+    private void handleRelease(String taskId) throws WorkflowEngineServerException {
```

Comment:
Same apply for this method. https://github.com/wso2-extensions/identity-workflow/pull/13/files#r2188997886